Add access token support (#71)
nginxinc/nginx-openid-connect: README.md

This implementation assumes the following environment:

> - The identity provider (IdP) supports OpenID Connect 1.0

Author: Yes it is automatically changed by Markdown Formatter. Addressed this with a [separate PR](https://github.com/nginxinc/nginx-openid-connect/pull/79).

- **openid_connect.server_conf** - this is the NGINX configuration for handling the various stages of OpenID Connect authorization code flow

  - No changes are usually required here
  - Modify the `resolver` directive to match a DNS server that is capable of resolving the IdP defined in `$oidc_token_endpoint`
  - If using [`auth_jwt_key_request`](http://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt_key_request) to automatically fetch the JWK file from the IdP then modify the validity period and other caching options to suit your IdP

- **openid_connect.js** - this is the JavaScript code for performing the authorization code exchange and nonce hashing
  - No changes are required unless modifying the code exchange or validation process

### Configuring the Key-Value Store

The key-value store is used to maintain persistent storage for ID tokens and refresh tokens. The default configuration should be reviewed so that it suits the environment. This is part of the advanced configuration in **openid_connect_configuration.conf**.

```nginx
keyval_zone zone=oidc_id_tokens:1M     state=conf.d/oidc_id_tokens.json     timeout=1h;

> keyval_zone zone=oidc_access_tokens:1M state=conf.d/oidc_access_tokens.json timeout=1h;

Author: Thanks for the great comments. Addressed it. Please let me know if I missed anything, or if you need more updates.

proxy_ssl_name        <IdP hostname>;
```

  * **Invalid access token**
    * The looping can now happen if the access token is not valid for some reason such that the setup of ID/access tokens expiration is different in the IdP.
    * In case of this, the backend/upstream can respond with HTTP 401 "Invalid token", and the frontend can restart the flow of OpenID Connect to get new ID token or refresh tokens by calling the following directives.

> ```nginx

Author: Removed it.

proxy_ssl_name        <IdP hostname>;
```

  * **Invalid access token**

>     Users may receive the response with HTTP 401 "invalid token" despite of successful authentication. There are several reasons why an OIDC access token might be not accepted by upstream although if it is not expired as follows:

Reviewer: ```suggestion
    Users may receive a `401` response with an optional "Invalid token" message despite successful authentication. There are several reasons why an OIDC access token might not be accepted by the upstream server, even if it has not expired:
```

>     * Incorrect backend server configuration. NGINX Plus sends bearer token in the HTTP Authorization header, but the backend application expects it in a specific cookie.

Reviewer: ```suggestion
    * Incorrect backend server configuration. NGINX Plus sends the bearer token in the HTTP Authorization header, but the backend application expects it in a specific cookie.
```

Author: Thanks! Revised!

    * The token has been tampered with. OIDC access tokens are digitally signed by the authorization server to ensure their authenticity. If the token has been modified in any way, the signature will no longer be valid, and the token will be considered invalid.

>     * It is important to note that the scope of an OIDC access token is independent of its validity. Even if an OIDC access token is not expired and has not been revoked, it may still be considered invalid if it does not have the necessary scope for the requested action. Please check "$oidc_scopes" variable in the "openid_connect_configuration.conf" file.

Reviewer: ```suggestion
    > **Note:** The scope of an OIDC access token is independent of its validity. Even if an OIDC access token is not expired and has not been revoked, it may still be considered invalid if it does not have the necessary scope for the requested action. Please check the `$oidc_scopes` variable in the `openid_connect_configuration.conf` file.
```

Author: Thanks! Revised!
